src/vm.rs

```diff
@@ -1941,6 +1954,141 @@ impl<'a, T: GuestStagePagingMode> FinalizedVm<'a, T> {
         Ok(0)
     }
 
+    fn rebind_vcpu_begin(&self, vcpu_id: u64, interrupt_file: ImsicFileId) -> EcallResult<()> {
+        self.vm()
+            .vcpus
+            .get_vcpu(vcpu_id)
+            .and_then(|vcpu| vcpu.rebind_imsic_prepare(interrupt_file))
+            .map_err(|_| EcallError::Sbi(SbiError::InvalidParam))
+    }
+
+    fn guest_rebind_vcpu_begin(
+        &self,
+        guest_id: u64,
+        vcpu_id: u64,
+        imsic_mask: u64,
+        active_vcpu: &ActiveVmCpu<T>,
+    ) -> EcallResult<u64> {
+        let guest = self.guest_by_id(guest_id)?;
+        let guest_vm = guest
+            .as_finalized_vm()
+            .ok_or(EcallError::Sbi(SbiError::InvalidParam))?;
+
+        // imsic_mask is in the same format as HGEIE, where bits [1:N] specify the guest interrupt
+        // files. We only support binding a single interrupt file for now.
+        if imsic_mask.count_ones() != 1 {
+            return Err(EcallError::Sbi(SbiError::InvalidParam))?;
+        }
+        let imsic_index = imsic_mask
+            .trailing_zeros()
+            .checked_sub(1)
+            .ok_or(EcallError::Sbi(SbiError::InvalidParam))?;
+
+        // Get the IMSIC page that we're going to assign.
+        let base_location = active_vcpu
+            .get_imsic_location()
+            .ok_or(EcallError::Sbi(SbiError::NotSupported))?;
+        let src_location = ImsicLocation::new(
+            base_location.group(),
+            base_location.hart(),
+            ImsicFileId::guest(imsic_index),
+        );
+        let from_page_addr = self
+            .vm_pages()
+            .imsic_geometry()
+            .and_then(|g| g.location_to_addr(src_location))
+            .ok_or(EcallError::Sbi(SbiError::InvalidParam))?;
+        let mut imsic_pages = self
+            .vm_pages()
+            .get_converted_imsic(from_page_addr)
+            .map_err(EcallError::from)?;
+
+        // to_page_addr contains the virtual address set by host. it's where guest vcpu views
+        // its interrupt file.
+        let to_page_addr = guest_vm.get_vcpu_imsic_addr(vcpu_id)?;
+
+        // Get the destination interrupt file.
+        //
+        // Unwrap ok: imsic_pages is exactly one page long and its location must be valid.
+        let interrupt_file = Imsic::get()
+            .phys_geometry()
+            .addr_to_location(imsic_pages.peek().unwrap())
+            .unwrap()
+            .file();
+
+        // Clears the new guest interrupt file and sets the state to Rebinding.
+        guest_vm.rebind_vcpu_begin(vcpu_id, interrupt_file)?;
+
+        // Change the virtual address mapping in second stage page table to now point to
+        // new physical address.
```

Comment:
Need comment about why `unwrap()` is ok here.

Reply:
Code no longer present.